### PR TITLE
[EC-286] Investigate use of DataSource on GenesisDataLoader

### PR DESCRIPTION
**Pull request by adamsmo**
This PR removes `DataSource` from `GenesisDataLoader` which is replaced by usage of `Blockchain`

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/blockchain/data/GenesisDataLoader.scala [hunk 1]
```diff
@@ -134,15 +118,33 @@ class GenesisDataLoader(
         Failure(new RuntimeException("Genesis data present in the database does not match genesis block from file." +
           " Use different directory for running private blockchains."))
       case None =>
-        // using empty namespace because ephemDataSource.storage already has the namespace-prefixed keys
-        ephemDataSource.storage.toSeq.grouped(dbConfig.batchSize).foreach(toStore => dataSource.update(IndexedSeq(), Nil, toStore))
+        // key.tail to remove namespace-prefix keys
```
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/blockchain/data/GenesisDataLoader.scala]: there should be option to remove prefix from key so GenesisDataLoader.scala does not depend on NodeStorage

File: src/main/scala/io/iohk/ethereum/blockchain/data/GenesisDataLoader.scala [hunk 2]
```diff
@@ -134,15 +118,33 @@ class GenesisDataLoader(
         Failure(new RuntimeException("Genesis data present in the database does not match genesis block from file." +
           " Use different directory for running private blockchains."))
       case None =>
-        // using empty namespace because ephemDataSource.storage already has the namespace-prefixed keys
-        ephemDataSource.storage.toSeq.grouped(dbConfig.batchSize).foreach(toStore => dataSource.update(IndexedSeq(), Nil, toStore))
+        ephemDataSource.getAll(nodeStorage.namespace)
+          .foreach { case (key, value) => blockchain.saveNode(ByteString(key.toArray[Byte]), value.toArray[Byte], 0) }
```
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/data/GenesisDataLoader.scala]: as far as I remember that grouping was added to (significantly?) improve performance. Don't we need it anymore?

Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/blockchain/data/GenesisDataLoader.scala]: good catch, will add it back

**Comment by adamsmo** [comment]
I tested performance on MacOs and it looks like version without direct access to data source is slower
original version loads genesis data in 5 seconds
version without direct access to data source loads genesis data in 7 seconds
there is no difference if we use batching or not in newer version
I was testing loading main net genesis file

**Comment by adamsmo** [comment]
for some reason this is failing when doing regular sync

**Comment by LukasGasior1** [comment]
is it failing only on regular sync? how so?

**Comment by adamsmo** [comment]
We are checking MPT state root because we are executing blocks

**Comment by AlanVerbner** [comment]
@adamsmo havent checks the code...but...be sure you are storing data as pruning mode would expect it (if basic pruning is enabled references count is stored alongside with mptnode)

**Comment by adamsmo** [comment]
I think issue is with loading genesis data through `blockchain.saveNode` with enabled pruning
I think it results in incorrect MPT state root hash

**Comment by LukasGasior1** [comment]
oh, I see. I thought it's the actual genesis init code that's failing

**Comment by LukasGasior1** [comment]
just a random guess: maybe something with prefixes (doubled or missing)?

**Comment by adamsmo** [comment]
@AlanVerbner good point, I think this is the issue that reference count is added twice, I am not removing this info when accessing mpt nodes direct in data source

**Comment by AlanVerbner** [comment]
Well....storages are dealing with serialization of data so if writing directly through the data source you need to be careful with that. I would avoid using data source directly

**Comment by adamsmo** [comment]
I think it is fixed, now since I am saving nodes with `blockchain.saveNode` I do not need to know what pruning is set for the app and I can set my internal pruning for checking MPT root hash to Archive which will not add any data that I will have to remove later

**Pull request merged**